adagrams/game.py

```diff
@@ -1,11 +1,89 @@
+from random import choices
+
+LETTER_POOL = {
+    "A": 9,
+    "B": 2,
+    "C": 2,
+    "D": 4,
+    "E": 12,
+    "F": 2,
+    "G": 3,
+    "H": 2,
+    "I": 9,
+    "J": 1,
+    "K": 1,
+    "L": 4,
+    "M": 2,
+    "N": 6,
+    "O": 8,
+    "P": 2,
+    "Q": 1,
+    "R": 6,
+    "S": 4,
+    "T": 6,
+    "U": 4,
+    "V": 2,
+    "W": 2,
+    "X": 1,
+    "Y": 2,
+    "Z": 1,
+}
+ONLY_LETTERS = [letter for letter in LETTER_POOL.keys()]
+ONLY_WEIGHTS = [num for num in LETTER_POOL.values()]
+LETTER_VALUE = {
+    1: ["A", "E", "I", "O", "U", "L", "N", "R", "S", "T"],
+    2: ["D", "G"],
+    3: ["B", "C", "M", "P"],
+    4: ["F", "H", "V", "W", "Y"],
+    5: ["K"],
+    8: ["J", "X"],
+    10: ["Q", "Z"],
+}
+
+
 def draw_letters():
-    pass
+    ten_letters = ""
+    while len(ten_letters) < 10:
+        letter = choices(ONLY_LETTERS, ONLY_WEIGHTS, k=1)[0]
+        if ten_letters.count(letter) < LETTER_POOL[letter]:
+            ten_letters += letter
+    return ten_letters
+
 
 def uses_available_letters(word, letter_bank):
-    pass
+    letter_bank_count = {}
+    for letter in letter_bank:
+        letter_bank_count[letter] = letter_bank.count(letter)
+    for letter in word.upper():
+        if letter in letter_bank and letter_bank_count[letter] > 0:
+            letter_bank_count[letter] -= 1
+        else:
+            return False
+    return True
+
 
 def score_word(word):
-    pass
+    total_score = 0
+    for element in word.upper():
+        for value, letters in LETTER_VALUE.items():
+            if element in letters:
+                total_score += value
```

Comment:
[optimization] You can `break` at this point, since none of the remaining `letters` will have `element`.

```suggestion
                total_score += value
                break
```

Reply:
[comment] Checking each letter in a range like this is definitely more maintainable than checking for each letter one by one. It could be faster, though. Consider how many checks this code does if `element` happens to be `"Z"`. How many more than `"A"`? If performance became a concern for this code, you could consider using a dictionary to map each possible letter to a score.

Building that dictionary is a little painful (and redundant), but we could imagine starting from a dictionary that uses a string of all the letters which share a score as the key, and the score as the value. That would be very maintainable and have no redundancy. Then we could write code to convert from that structure to a dictionary where each letter is its own key. This would let us have a maintainable data representation, AND quick score lookups!

That might look like starting with your dictionary and adding a function to build a per-letter dictionary of scores...

```python
LETTER_VALUE = {
    1: ["A", "E", "I", "O", "U", "L", "N", "R", "S", "T"],
    # ...
    10: ["Q", "Z"],
}

def build_letter_to_score():
  """ produces a dictionary of letter to score, e.g. { "A": 1, "E": 1, ... "Z": 10 } """
  ...
```

You will often find that, on the job, this kind of extra effort is worth the development time to balance performance and maintainability.